internal/mesh/internal/controllers/sidecarproxy/builder/local_app.go

```diff
@@ -0,0 +1,141 @@
+package builder
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/envoyextensions/xdscommon"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1/pbproxystate"
+)
+
+func (b *Builder) BuildLocalApp(workload *pbcatalog.Workload) *Builder {
+	return b.addInboundListener(xdscommon.PublicListenerName, workload).
+		addInboundRouters(workload).
+		addInboundTLS()
+}
+
+func (b *Builder) getLastBuiltListener() *pbproxystate.Listener {
+	lastBuiltIndex := len(b.proxyStateTemplate.ProxyState.Listeners) - 1
+	return b.proxyStateTemplate.ProxyState.Listeners[lastBuiltIndex]
+}
+
+func (b *Builder) addInboundListener(name string, workload *pbcatalog.Workload) *Builder {
+	listener := &pbproxystate.Listener{
+		Name:      name,
+		Direction: pbproxystate.Direction_DIRECTION_INBOUND,
+	}
+
+	// We will take listener bind port from the workload for now.
+	// Find mesh port.
+	var meshPort string
+	for portName, port := range workload.Ports {
+		if port.Protocol == pbcatalog.Protocol_PROTOCOL_MESH {
+			meshPort = portName
+			break
+		}
+	}
+
+	// Check if the workload has a specific address for the mesh port.
+	var meshAddress string
+	for _, address := range workload.Addresses {
+		for _, port := range address.Ports {
+			if port == meshPort {
+				meshAddress = address.Host
```

Comment:
What if there are multiple addresses? You don't break out of both loops, so each time this iterates it will ultimately choose the last value. Is that the value we want?

Reply:
great catch! I updated it to not do that. I think ultimately we want to allow only one non-external mesh address for the workload so I added that as a separate task in Jira.